OpenRA.Game/FieldLoader.cs

```diff
@@ -174,396 +627,44 @@ public static object GetValue(string fieldName, Type fieldType, string value, Me
 		public static object GetValue(string fieldName, Type fieldType, MiniYaml yaml, MemberInfo field)
 		{
 			var value = yaml.Value?.Trim();
-
-			if (fieldType == typeof(int))
-			{
-				if (Exts.TryParseIntegerInvariant(value, out var res))
-					return res;
-				return InvalidValueAction(value, fieldType, fieldName);
-			}
-			else if (fieldType == typeof(ushort))
-			{
-				if (ushort.TryParse(value, NumberStyles.Integer, NumberFormatInfo.InvariantInfo, out var res))
-					return res;
-				return InvalidValueAction(value, fieldType, fieldName);
-			}
-
-			if (fieldType == typeof(long))
-			{
-				if (long.TryParse(value, NumberStyles.Integer, NumberFormatInfo.InvariantInfo, out var res))
-					return res;
-				return InvalidValueAction(value, fieldType, fieldName);
-			}
-			else if (fieldType == typeof(float))
-			{
-				if (value != null && float.TryParse(value.Replace("%", ""), NumberStyles.Float, NumberFormatInfo.InvariantInfo, out var res))
-					return res * (value.Contains('%') ? 0.01f : 1f);
-				return InvalidValueAction(value, fieldType, fieldName);
-			}
-			else if (fieldType == typeof(decimal))
-			{
-				if (value != null && decimal.TryParse(value.Replace("%", ""), NumberStyles.Float, NumberFormatInfo.InvariantInfo, out var res))
-					return res * (value.Contains('%') ? 0.01m : 1m);
-				return InvalidValueAction(value, fieldType, fieldName);
-			}
-			else if (fieldType == typeof(string))
-				return value;
-			else if (fieldType == typeof(Color))
-			{
-				if (value != null && Color.TryParse(value, out var color))
-					return color;
-
-				return InvalidValueAction(value, fieldType, fieldName);
-			}
-			else if (fieldType == typeof(Hotkey))
-			{
-				if (Hotkey.TryParse(value, out var res))
-					return res;
-
-				return InvalidValueAction(value, fieldType, fieldName);
-			}
-			else if (fieldType == typeof(HotkeyReference))
+			if (fieldType.IsGenericType)
 			{
-				return Game.ModData.Hotkeys[value];
-			}
-			else if (fieldType == typeof(WDist))
-			{
-				if (WDist.TryParse(value, out var res))
-					return res;
-
-				return InvalidValueAction(value, fieldType, fieldName);
-			}
-			else if (fieldType == typeof(WVec))
-			{
-				if (value != null)
-				{
-					var parts = value.Split(',');
-					if (parts.Length == 3)
-					{
-						if (WDist.TryParse(parts[0], out var rx) && WDist.TryParse(parts[1], out var ry) && WDist.TryParse(parts[2], out var rz))
-							return new WVec(rx, ry, rz);
-					}
-				}
-
-				return InvalidValueAction(value, fieldType, fieldName);
+				if (GenericTypeParsers.TryGetValue(fieldType.GetGenericTypeDefinition(), out var parseFuncGeneric))
+					return parseFuncGeneric(fieldName, fieldType, yaml, field);
```

Comment:
Was there a specific reason to not pass `value` here like you've done for the other cases?
This would allow the  duplicated `var value = yaml.Value?.Trim();` in the generic `Parse*` methods to be removed.

Reply:
I believe at least one of the generic type parser functions uses the yaml. The Dictionary.

Reply:
How about passing both?